feat: asynchronously import axe-core

raunofreiberg opened the pull request:
Closes #9

<hr />

## Benchmarks

I used the default CRA template and wrapped the application as:

```tsx
<AxeMode disabled={process.env.NODE_ENV !== 'development'}>
```

And everything related to `axe-mode` was put into a separate chunk, and dropped in production 🎉:

![image](https://user-images.githubusercontent.com/23662329/81328354-682cee00-90a5-11ea-8642-608c55f4ea2e.png)

chancestrickland commented on src/index.tsx:
```diff
@@ -210,7 +210,9 @@ export default function AxeMode({ children, disabled }: AxeModeProps) {
   const childrenRef = React.useRef<HTMLElement | null>(null);
 
   React.useEffect(() => {
-    import('axe-core').then(setAxe);
```
I think it's probably OK to import the module regardless of whether or not it's disabled. You only want to do it once anyway, and that way if the user toggles from disabled after the component is mounted it should be ready to go (not that it would be that slow anyway, but still).

So I'd probably just check to make sure the component is still mounted in the async callback and leave out the disabled check.

```js
  React.useEffect(() => {
    let mounted = true;
    import('axe-core').then((axe) => {
      if (mounted) {
        setAxe(axe);
      }
    });
    return () => {
      mounted = false;
    }
  }, []);
```

raunofreiberg replied:
Hmm. I think we should check for `disabled` before importing the module.

Webpack will split `axe-mode` into a separate chunk since it is imported dynamically. Although, if `AxeMode` is wrapped around the application at all times, we should make sure not to import the dynamic module when it's disabled, which is done in user-land by setting `disabled={process.env.NODE_ENV !== 'development'`. Otherwise, we will always import the chunk and essentially this does not have the desired effect of not requesting it in production.

chancestrickland replied:
It should probably be disabled completely in production anyway IMO. Maybe we're trying to optimize something a bit too early here, I'm not sure the bundle will be a bottleneck for development.

```js
const AxeModeImpl = () => () // real implementation
export const AxeMode = process.env.NODE_ENV !== 'development' ? React.Fragment : AxeModeImpl
```

raunofreiberg replied:
Totally agree!

I think in your example, `AxeMode` would have to be a component that imports `AxeModeImpl` based on `NODE_ENV`. Otherwise, it will be a top-level import and bundled alongside the application code.

I think we can import `axe-core` synchronously then?

chancestrickland replied:
@raunofreiberg Yeah I think that's a somewhat better approach. I'll add to the PR.

chancestrickland commented:
@raunofreiberg Check out my latest commits and let me know what you think. GitHub won't diff correctly because I renamed the implementation file, but all I changed there was using a standard sync import for `axe-core` and lazy loading our implementation in `index.ts`.

chancestrickland commented:
BTW, had some issues with the build and it looks like TSDX doesn't handle lazy loading at the moment. I updated the config but now we're not getting CJS bundles. This needs some work but we can probably open another PR for that.

raunofreiberg commented:
@chancestrickland Great work!

I modified the entry point a bit. It did not seem to work otherwise with a production build.

The entry point has to be a component that checks for `NODE_ENV` since there needs to be a `React.Suspense` around the lazy-loaded component. 

I tested it out on the examples and a CRA template and the whole implementation of our module is split into a separate chunk, and dropped in production!

Lemme know if this makes sense!
